src/Tools/AnalyzerRunner/Program.cs

```diff
@@ -0,0 +1,480 @@
+                }
+            }
+        }
+
+        private static async Task<ImmutableDictionary<ProjectId, ImmutableArray<Diagnostic>>> GetAnalyzerDiagnosticsAsync(Solution solution, string solutionPath, ImmutableArray<DiagnosticAnalyzer> analyzers, Options options, TelemetryCollector telemetryCollector, CancellationToken cancellationToken)
+        {
+            List<KeyValuePair<ProjectId, Task<ImmutableArray<Diagnostic>>>> projectDiagnosticTasks = new List<KeyValuePair<ProjectId, Task<ImmutableArray<Diagnostic>>>>();
+
+            // Make sure we analyze the projects in parallel
+            foreach (var project in solution.Projects)
+            {
+                if (project.Language != LanguageNames.CSharp)
+                {
+                    continue;
+                }
+
+                projectDiagnosticTasks.Add(new KeyValuePair<ProjectId, Task<ImmutableArray<Diagnostic>>>(project.Id, GetProjectAnalyzerDiagnosticsAsync(analyzers, project, options, telemetryCollector, cancellationToken)));
+            }
+
+            ImmutableDictionary<ProjectId, ImmutableArray<Diagnostic>>.Builder projectDiagnosticBuilder = ImmutableDictionary.CreateBuilder<ProjectId, ImmutableArray<Diagnostic>>();
+            foreach (var task in projectDiagnosticTasks)
+            {
+                projectDiagnosticBuilder.Add(task.Key, await task.Value.ConfigureAwait(false));
+            }
+
+            return projectDiagnosticBuilder.ToImmutable();
+        }
+
+        /// <summary>
+        /// Returns a list of all analyzer diagnostics inside the specific project. This is an asynchronous operation.
+        /// </summary>
+        /// <param name="analyzers">The list of analyzers that should be used</param>
+        /// <param name="project">The project that should be analyzed</param>
+        /// <see langword="false"/> to use the behavior configured for the specified <paramref name="project"/>.</param>
+        /// <param name="cancellationToken">The cancellation token that the task will observe.</param>
+        /// <returns>A list of diagnostics inside the project</returns>
+        private static async Task<ImmutableArray<Diagnostic>> GetProjectAnalyzerDiagnosticsAsync(ImmutableArray<DiagnosticAnalyzer> analyzers, Project project, Options analyzerOptionsInternal, TelemetryCollector telemetryCollector, CancellationToken cancellationToken)
+        {
+            Utilities.WriteLine($"Running analyzers for {project.Name}", ConsoleColor.Gray);
+            var supportedDiagnosticsSpecificOptions = new Dictionary<string, ReportDiagnostic>();
+
+            // Report exceptions during the analysis process as errors
+            supportedDiagnosticsSpecificOptions.Add("AD0001", ReportDiagnostic.Error);
+
+            // update the project compilation options
+            var modifiedSpecificDiagnosticOptions = supportedDiagnosticsSpecificOptions.ToImmutableDictionary().SetItems(project.CompilationOptions.SpecificDiagnosticOptions);
```

Comment:
it could just create ImmutableDictionary from the beginging and do Add?

Reply:
Not sure I understand completely. We can create supportedDiagnosticsSpecificOptions in the very beginning and then SetItems over it.

modifiedSpecificDiagnosticOptions are created per project and just once. Do you see an option to reuse it?

---
In reply to: [149863323](https://github.com/dotnet/roslyn/pull/23087#discussion_r149863323) [](ancestors = 149863323)

Reply:
no what I meant was, it is creating regular dictionary and add "Ad001" and then convert to immutable dictionary and then merge it with compilationOptions.Specific...Options.

it could be just
var modified..Options = CompilationOptions.Specif...Options.Add("AD0001", ....) or CompilationOptions.Specific...Options["AD0001"] = ReportDiagnostic.Error.

that conversion is unnecessary.